forem/forem
# Admin member index: show/hide search and filter controls

## What type of PR is this? (check all applicable)


## Description

This PR addresses the final task on https://github.com/forem/forem/issues/16933. When viewing the main admin member view (`/admin/users`) on a smaller screen, you should see search and filter icons, rather than the form fields. Clicking each icon expands the relevant section, and all functionality remains the same.

To help users understand when an "invisible" field has a value in it, we have an indicator displayed. This is shown when the user has a value in a field, but the field is _not_ expanded (see video below).

On larger screens all that's changing here is some minor style tweaking, e.g. adding an additional search submit icon button over the search input. 

NB: This PR doesn't tackle adding the same controls to the "Invited" or "GDPR delete requests" views. Initially, I thought this could neatly fold in, but given those views aren't currently searchable right now and Phase 1 of this project doesn't include those views, it made sense to keep those concerns separate. I've created draft issues on the project board for them instead.

## Related Tickets & Documents



- Closes https://github.com/forem/forem/issues/16933

## QA Instructions, Screenshots, Recordings

With the `:member_index_view` feature flag disabled, everything should be exactly the same as before 😄 
But when it is enabled, you can test by:

- Visiting `admin/users` as an admin
- Try the search and filter by role functionality
- Resize your browser window until the search field/role select collapses and you see only a button for each
- Try toggling the expandable views by clicking on them
- Try toggling between the expandable views
- Try searching and filtering
- On a small screen size try adding values then collapsing sections etc, to check the indicator shows when appropriate (for example see video below)


https://user-images.githubusercontent.com/20773163/160422699-9b7867d8-def5-4545-ab9f-a6c78d80aeee.mp4


### UI accessibility concerns?

- I've added `aria-expanded` to the buttons to make sure their purpose and state is communicated
- When a section is expanded, I'm sending focus directly to the search/select input. Otherwise a keyboard user would have to tab through the pagination to get there, and a screen reader user would have to search around for the new content

On the slightly negative side, we've added a second form submit button in large screen layout, as we now have:
- the filter button
- the search button

Both buttons submit the same form (i.e. they both search by name _and_ filter), so it's perhaps not ideal that a keyboard user now has an extra tab stop added. However, on balance I thought it best to add it in, since it keeps it consistent with other views (e.g. Invitations) which don't have the "filter by role" option, and therefore require the explicit search button. 

## Added/updated tests?



## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [X] This change does not need to be communicated, and this is why not: Behind a feature flag as part of a larger epic

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![cute](https://media.giphy.com/media/l0Ex33YvFIfFJU0uc/giphy.gif)
